### ISSUE TITLE
"unevaluatedItems" should consider "contains"

### ISSUE THREAD

**Issue by handrews** [json-schema-org/json-schema-spec#810]
The interaction of these seems underspecified.  @Relequestual what was the problem you ran into here exactly?  I've already forgotten 😬

**Comment by Relequestual** [comment]
This was me being unable to read that unevaluatedItems only applies when items is defined as an array.

With that light, there’s no valid argument that I can think of that would warrant interaction between these keywords.

**Comment by handrews** [comment]
Actually, I'm pretty sure I left out something important.  Consider the following schema:

```JSON
{
  "type": "array",
  "oneOf": [
    {
      "contains": {"type": "string"},
    },
    {
      "items": {"type": "integer"}
    }
  ],
  "unevaluatedItems": {"type": "boolean"}
}```

The following instances should all be valid:

`[true, "hello", false]`
`[1, 3, 4]`
`[true, false]`

I had intended for all array child applicators to affect `unevaluatedItems`, and all object child value applicators to affect `unevaluatedProperties`.  (`propertyNames` is an object child name applicator).

I seem to have left out the necessary specification of a `contains` annotation and how `unevaluatedItems` could use it.

Not sure if there's a use case, but logical consistency was the goal.
@Relequestual does this make sense?

**Comment by Relequestual** [comment]
Logical.
Would contains collect all items that were valid for it? As such, would unevaluatedItems only apply to those which were not valid according to contains?

Sorry, on mobile.

**Comment by handrews** [comment]
@Relequestual 

> Would contains collect all items that were valid for it?

Yes, there is language in the existing description that requires `contains` to be checked against all array items when collecting annotations.  If you're using `unevaluatedItems` you have to collect annotations.  I'll make sure that's clear, though, good question.

> would unevaluatedItems only apply to those which were not valid according to contains?

That is correct.  `contains` gets applied to all but "evaluated" really means "successfully evaluated" because annotations are dropped from failed evaluations.

**Comment by handrews** [comment]
Hm... lots of test cases there.

**Comment by Relequestual** [comment]
Well, my understanding / expectation is correct then.
I'd have to re-read the spec to understand where you're saying there's a deficit though.

**Comment by handrews** [comment]
@Relequestual if you find anywhere that makes that behavior explicit, let me know.

I don't think there's anything there that contradicts it, and it obviously won't require a meta-schema update.  I made that "draft-08-patch1" milestone to collect compatible fixes based on early feedback.  Or if we end up thinking that this is a kind of glaring error we can put this and whatever else we've noticed in and re-publish the core spec sooner rather than later.

**Comment by Relequestual** [comment]
I think the phrase "annotations are dropped from failed evaluations" is key here.

Does `contains` return an array of annotations with nulls where the annotations have been dropped (becaused of failed validation), or an object where the keys are the indexes at which the array successfully validated? Latter sounds better. I don't think it's currently defined.

It only really matters what format the annotations take for the output format data.

I guess `contains` is an applicator keyword like `anyOf`. At least the intent is clearer and should help people avoid using `oneOf` here... =D

The prose / explanation about how annotation results are used will need to be updated here: https://github.com/json-schema-org/json-schema-spec/blob/38e1606d71ee848c6cf0490c3457681220c01663/jsonschema-core.xml#L2209

**Comment by Relequestual** [comment]
Given the proceeding paragraphs, making this fix would legalistically be a non backwards compatible change, but I think it's clearly a mistake.

https://github.com/json-schema-org/json-schema-spec/blob/38e1606d71ee848c6cf0490c3457681220c01663/jsonschema-core.xml#L2219-L2232

**Comment by handrews** [comment]
@Relequestual yeah, this falls under "bugfix"

**Comment by gregsdennis** [comment]
@handrews I don't think that `[true, false]` should pass: it doesn't meet the `oneOf` requirement.

**Comment by gregsdennis** [comment]
>>would unevaluatedItems only apply to those which were not valid according to contains?
>
>That is correct. contains gets applied to all but "evaluated" really means "successfully evaluated" because annotations are dropped from failed evaluations.

Also, I don't think I'm doing this right.

**Comment by handrews** [comment]
@gregsdennis 

> @handrews I don't think that [true, false] should pass: it doesn't meet the oneOf requirement. For this case, unevaluatedItems is irrelevant.

Yes, that's correct, good catch.

**Comment by Relequestual** [comment]
I read that and made a wrong assumption. Yikes.
Let me fish out my example from my slides for this. It's a tricky one to reason about.

**Comment by Relequestual** [comment]
This issue has changed from it's original, but that's not a problem, as a bug is still highlighted.
I'll modify the title and first comment to reflect.

**Comment by Relequestual** [comment]
Wait, no. We are forgetting something here.

`additionalItems` is only applicable when `items` is an array, not when `items` is a single schema.
SO, the example @handrews used above (https://github.com/json-schema-org/json-schema-spec/issues/810#issuecomment-544216155) wouldn't be useful for `additionalItems`, and the logic behind `unevaluatedItems` relies on annotation collection from `items` as an array only.

From draft-7...


https://github.com/json-schema-org/json-schema-spec/blob/1afc34b65ead445ff363cfc870a28de0ca56e20f/jsonschema-validation.xml#L527-L536

If we want to change this, so you COULD do `additionalItems` or `unevaluatedItems` where `items` isn't an array, then the annotation collection needs to be changed to not just track the max index to which the subschema value of `items` was applied, but the result of applying it to every item in the array it's being applied to.

Thoughts?

**Comment by Relequestual** [comment]
I'm opening a channel for this on slack as I think we will have some discussion. We should make sure we are on the same page and understand correctly before determining what to do here.

**Comment by handrews** [comment]
`contains` should just return a list of indices that validated as its annotation.  That's the minimum amount of information needed to use `unevaluatedItems`.  And if array `items` and `contains` were both used, then an implementation can just look for indices greater than the highest index from the `items` array.

**Comment by Relequestual** [comment]
@handrews @gregsdennis I think I follow now.
Here are some examples of how the annotation collection for `contains` should work and how it relates to `unevaluatedItems`


```jsonc

{
  "type": "array",
  "oneOf": [
    {
      "contains": {"type": "string"}
    },
    {
      "items": {"type": "integer"}
    }
  ],
  "unevaluatedItems": {"type": "boolean"}
}

// Valid Instances

  [1, 3, 4 ],
  // each item is valid according to the `items` schema value which results in a `true` annotation, 
  // so `unevaluatedItems` is not applied

  ["a", "b", "c"],
  // each item is valid when the `contains` schema value is applied, resulting in an annotation of [1, 2, 3],
  // there are no additional items in the array, so `unevaluatedItems` is not applied

  [true, "hello", false],
  // `items` fails validation, so produces no annotation
  // `contains` is applied, and results in an annotation of [1] for the index of which item is valid
  // `unevaluatedItems` is applied to the other indexs of the array

// Invalid instances 

  ["a", false, 2, "b", 8 ]
  // `items` is not met, so produces no annotation
  // `contains` is applied and results in an annotation of [0, 3]
  // `unevaluatedItems` is applied to the other items in the array, and fails because numbers are not booleans


  // Another example
{
  "allOf": [
    {
      "items": [
        { "type": "integer" },
        { "type": "string" }
        true,
      ]
    },
    {
      "contains": { "type": "object", "required": ["error"] }
    }
  ],
  "unevaluatedItems": { "type": "object", "required": ["debug"] }

}

// Valid

  [1, "errorABC", { "error": 123 }]
  // Int, string, anything is fine
  // Must contain an object with an error
  // `items` results in annotation of [0,1,2]
  // `contains` results in annotation of [2]
  // `unevaluatedItems` is not applied to the array as there are not additional items beyond item at index 2.

  [2, "errorABC", "fail reason as string", { "error": 123 }]
  // `items` results in annotation of [0,1,2]
  // `contains` results in annotation of [3]
  // `unevaluatedItems` is not applied to the array as above

  [3, "errorABC", "fail reason as string", { "error": 123 }, { "error": 456 }]
  // `items` results in annotation of [0,1,2]
  // `contains` results in annotation of [3, 4]
  // `unevaluatedItems` is not applied to the array as above

  [4, "errorABC", "fail reason as string", { "error": 123 }, { "debug": "some debug info" }]
  // Same annotations as previous
  // item at index 4 has `unevaluatedItems` schema value applied

  [5, "errorABC", "fail reason as string", { "debug": "some debug info" }, { "error": 123 }]
  // `items` results in annotation of [0,1,2]
  // `contains` results in annotation of [4]
  // `unevaluatedItems` is applied to item at index 3

// Invalid

  [6, "errorABC", "fail reason as string", { "info": "some other info" }, { "error": 123 }]
  // `items` results in annotation of [0,1,2]
  // `contains` results in annotation of [4]
  // `unevaluatedItems` is applied to item at index 3
  // Validation fails for item at index 3 when `unevaluatedItems` schema value is applied.

  [7, "errorABC", "fail reason as string", false, { "error": 123 }]
  // Same as above

  [8, "errorABC", "fail reason as string", { "error": 123 }, { "info": "some other info" }, { "error": 456 }]
  // `items` results in annotation of [0,1,2]
  // `contains` results in annotation of [3, 5]
  // `unevaluatedItems` is applied to item at index 4
  // Validation fails for item at index 4 when `unevaluatedItems` schema value is applied.

```

**Comment by gregsdennis** [comment]
I think that lines up with what @handrews is saying, but I don't agree with `unevaluatedItems` not covering items that the `contains` subschema matches on.

In my interpretation, `contains` applies not to the item, but to the array.  While it validates that one of the items matches its subschema, its annotations should apply to the *array*, not the item that passes.

That means that your third case changes:

```
  [true, "hello", false],
  // `items` fails validation, so produces no annotation
  // `contains` is applied, and returns a result of true on the array because the array contains a matching item (the index of the item is 1, and this can be an annotation)
  // `unevaluatedItems` is applied to all indexes of the array because the items themselves haven't been annotated.
```

---

Writing that out, it feels wrong.  Maybe I'm getting hung up on the word "contains" meaning that its target is the array and is validating whether it contains something.  Instead it's working as sort of a "some of the items should be this, and I want to know which items they are, then don't re-evaluate them, like ever."  But I don't know of a single word that does that, so maybe "contains" works, but it still feels weird.

In short, @handrews wrote the `unevaluatedItems` stuff so he knows best how he intended it to work.  Go with what he says.

**Comment by gregsdennis** [comment]
@handrews another aspect to this: how does `contains` fit into the evaluation sequence?  For example, we do `items` -> `additionalItems` -> `unevaluatedItems`.  `unevalutatedItems` is obviously last, but where does `contains` fit in with the rest?

**Comment by jdesrosiers** [comment]
Here's another perspective to consider.

The array form of `items` is a hack to support a tuple data structure. It only uses `type` "array" because there isn't a "tuple" JSON type. The `additionalItems` keyword is only defined for use with an array-form-`items` keyword. In other words, it only applies to tuples, not arrays. I would expect the `unevaluatedItems` keyword to also only apply to tuples. Therefore, the `items` and `contains` keywords would not affect `unevaluatedItems`.

Consider the following example. We wouldn't expect an item that matches `contains` to be exempt from `addtionalItems`, right? `additionalItems` applies to any item in addition to what is defined in an array-form-`items` keyword. If array-form-`items` isn't present, `additionalItems` applies to everything. An instance validated against this schema would always fail because either "hi" is missing or "hi" has too few characters.
```json
{
  "type": "array",
  "items": [{ "const": "foo" }, { "const": "bar" }, { "const": "baz" }],
  "additionalItems": { "type": "string", "minLength": 3 },
  "contains": { "const": "hi" }
}
```

I think `unevaluatedItems` should work the same way. It's for when you want to use `additionalItems` with an extended tuple. Just like `unevaluatedProperties` is for when you want to use `additionalProperties` with an extended object.
```json
{
  "type": "array",
  "allOf": [
    { "items": [{ "const": "foo" }] },
    { "items": [true, { "const": "bar" }] }
  ],
  "unevaluatedItems": false
}
```

**Comment by handrews** [comment]
@jdesrosiers 

> If array-form-items isn't present, `additionalItems` applies to everything. 

As explained in [section 9.3.1.2 "additionalItems"](https://tools.ietf.org/html/draft-handrews-json-schema-02#section-9.3.1.2) (emphasis added):

> Otherwise, if "items" **_is absent_** or its annotation result is the boolean true, "additionalItems" MUST be ignored.

Several drafts ago we discussed this fairly extensively and decided that there should only be one way apply a subschema to every element of an array, and that was the single-schema `items`.  While this is an asymmetry with the behavior of `additionalProperties` w.r.t. `properties`, that asymmetry was already present due to `properties` not having a single-schema form.

We considered several options, including giving `properties` a single-schema form, outright forbidding `additionalItems` to be present without `items`, and I think at least one other approach that escapes me at the moment.  We settled on the wording above and it seems to have held up fine since then.

------

I also don't follow your argument in general.  If we change your first example slightly:

```JSON
{
  "type": "array",
  "items": [
    { "type": "string", "minLength": 2},
    { "type": "string", "minLength": 2},
    { "const": "foo" }
  ],
  "additionalItems": { "type": "string", "minLength": 3 },
  "contains": { "const": "hi" }
}
```

Then now this can be satisfied by instances like
`["hi", "foo", "bar", "some", "longer", "strings"]`

The equivalent with `unevaluatedItems` would then be:

```JSON
{
  "type": "array",
  "allOf": [
    { "items": [{"minLength": 2}, { "const": "foo" }] },
    { "items": [true, true, { "const": "bar" }] },
    {"contains": {"const": "hi"}}
  ],
  "unevaluatedItems": false
}
```

**Comment by jdesrosiers** [comment]
> if "items" is absent [...], "additionalItems" MUST be ignored.

Thanks for the correction. I didn't realize this had changed. Since there's some confusion about the argument I'm making, I'll point out that this detail has nothing to do with my explanation.

You have a few issues with your examples, but I think I can read between the lines and say we agree that those two schemas are equivalent. I'm not sure what you're trying to point out here.

**Comment by Relequestual** [comment]
> In my interpretation, contains applies not to the item, but to the array. While it validates that one of the items matches its subschema, its annotations should apply to the array, not the item that passes. - @gregsdennis 

@handrews For me, this is enough justification for `additionalItems` and `unevaulatedItems` to NOT consider the annotation result of `contains`.

That being said, it may still be useful to collect the annotations which indicate WHICH items in the array pass the `contains` schema.

Assuming we resolve #864 and end up with `tupleItems` etc, then `unevaluatedITupleItems`, hopefully it will be clearer that there's no interaction with `contains.

Thoughts?

**Comment by handrews** [comment]
@Relequestual I disagree with @gregsdennis's interpretation.  `contains` applies the subschema to each item in turn, and the annotations from that subschema are most definitely attached _to the item_.  This is exactly the same as what `items` does.  The ***only*** difference between `items` and `contains` is that the former is an AND and the latter is an OR.

Let's pretend we had a keyword `singleItem` that takes a schema and an index of the item to which it applies.  It puts that index in its annotations results.  So if it successfully applies to index 2, then it has an annotation of 2, plus the subschema's annotations.

If we could write out an infinite sequence of these, so that they could cover an indefinitely long array, we could write `items` and `contains` in terms of this keyword.

`items` would be:
```JSON
"allOf": [
  {
    "singleItem": [0, {...}]
  },
  {
    "singleItem": [1, {...}]
  },
  {
    "singleItem": [2, {...}]
  },
  ...
]
```
`contains` would be:
```JSON
"anyOf": [
  {
    "singleItem": [0, {...}]
  },
  {
    "singleItem": [1, {...}]
  },
  {
    "singleItem": [2, {...}]
  },
  ...
]
```

So, if we said that `unevaluatedItems` does ***not*** apply to any index that `singleItem` returns as an annotation, _that_ is the desired behavior.

Is this more clear?

**Comment by jdesrosiers** [comment]
Conceptually, `unevaluatedItems` should work exactly like `additionalItems`, it just has the extra super power of being able to consider subschemas. Right? If `contains` can influence `unevaluatedItems`, it no longer works the same way as `additionalItems`. Here's an example.

```json
{
  "type": "array",
  "items": [{ "type": "string" }, { "type": "number" }],
  "contains": { "const": 42 },
  "additionalItems": { "type": "number", "maximum": 10 }
}
```

Example 1: `["foo", 42, 1]` => Valid
Example 2: `["foo", 1, 42]` => Invalid

If we replace `additionalItems` with `unevaluatedItems`, `unevaluatedItems`'s super power is not relevant because there are no subschemas to consider. Therefore, we should expect `unevaluatedItems` to behave exactly like `additionalItems` did in the previous schema.

```json
{
  "type": "array",
  "items": [{ "type": "string" }, { "type": "number" }],
  "contains": { "const": 42 },
  "unevaluatedItems": { "type": "number", "maximum": 10 }
}
```

*Assuming `contains` can impact `unevaluatedItems`*
Example 1: `["foo", 42, 1]` => Valid
Example 2: `["foo", 1, 42]` => **Valid**

Notice that the validation result for the second example has changed.

I strongly believe that nothing should change, but **if there's consensus that `contains` should impact `unevaluatedItems`, then `additionalItems` should be changed to match that behavior.**

**Comment by gregsdennis** [comment]
Core Specification Section 9.1:

> Schema keywords typically operate independently, without affecting each other's outcomes.
>
> For schema author convenience, there are some exceptions among the keywords in this vocabulary:
>
> - "additionalProperties", whose behavior is defined in terms of "properties" and "patternProperties"
> - "unevaluatedProperties", whose behavior is defined in terms of annotations from "properties", "patternProperties", "additionalProperties" and itself
> - "additionalItems", whose behavior is defined in terms of "items"
> - **"unevaluatedItems", whose behavior is defined in terms of annotations from "items", "additionalItems" and itself**

This makes no mention of `contains`.

Section 9.3.1.4:

> The value of this keyword MUST be a valid JSON Schema.
>
> **An array instance is valid against "contains" if at least one of its elements is valid against the given schema.** Note that when collecting annotations, the subschema MUST be applied to every array element even after the first match has been found. This is to ensure that all possible annotations are collected.

The *array* is valid, not the element.  The subschema must be applied to all elements when collecting annotations, but it *does not* say that the annotation applies to the element.  These two combined implies that the annotations apply to the array.

**Comment by Relequestual** [comment]
@gregsdennis I'm not quite sure why you're quoting the spec here. We're trying to work out if the spec IS right or needs changes.

@jdesrosiers I can see your point, and conceptually I agree, but in reality I think that IS OK.

After all, the keyword `unevaulatedItems` is looking to catch items which have not been evaluated and produced annotation results (aka have a pass assertion).


@handrews I think I agree with you here.
For the sake of clarity, can you confirm if you agree with my examples in my previous comments and how they are reasoned?

The example @jdesrosiers provided showing how the `*items` keywords work differently than one might conceptualize could prove valuable for adding to tests.

**Comment by handrews** [comment]
I probably agree with @jdesrosiers on

> I strongly believe that nothing should change, but if there's consensus that `contains` should impact `unevaluatedItems`, then `additionalItems` should be changed to match that behavior.

at least in theory.  Whether that means `unevaluatedItems` should align with `additionalItems` after all, or if `additionalItems` should define behavior w.r.t. `contains` is less clear to me.

I don't think anyone has ever asked how or if `additionalItems` and `contains` interact.  I don't recall giving it any thought at all.  Mostly because we didn't have a way to talk about how these keywords worked, they just each did stuff.

So, on the plus side, defining consistent mechanisms is helping us figure out how keywords should behave.  On the minus side, that's kind of irritating 😛

**Comment by gregsdennis** [comment]
@Relequestual I'm quoting the spec because my understanding is derived from reading it.  Ergo, if my understanding is wrong, then the spec needs to change to be more explicit about the intended meaning.

It's always helpful to know what the current state is when deciding whether it should change.

**Comment by Relequestual** [comment]
you are right @gregsdennis in your reading and understading of the spec 2019-09, but it also says the subschema is applied to each item in the array... but I get your point.

**Comment by Relequestual** [comment]
I'm marking this as blocked till #864 is resolved.

**Comment by Relequestual** [comment]
Possible use case: https://stackoverflow.com/questions/60724246/how-to-set-limit-to-json-schema-objects-value

**Comment by handrews** [comment]
Good use case from @rjmill 

> For example, it could be cool to ensure that the “contains” item is in the array, but that everything else validates against another schema.

**Comment by handrews** [comment]
We decided to go with the two-keyword `prefixItems` + `items`, which removes the concern over whether `contains` needed to impact `additionalItems` in order to parallel any impact on `unevaluatedItems`.

This further validates breaking the array/object symmetry by dumping `additionalItems`.  `items` is defined purely in terms of `prefixItems` (the items before `items`) rather than a more general question of what else has been validated (additional beyond anything else in the same schema object).  `unevaluatedItems` is a general "what has been (successfully) validated" keyword.

This means that `contains` + `unevaluatedItems` can be used for the "apply a subschema to every item that doesn't match `contains`" use case in the previous comment, while `contains` + `items` can be used for layering an additional `contains` check on top of a uniform schema for all (or all after `prefixItems`) positions.

I'm going to write a PR for both this and #864 as they need to rework basically the same areas of the spec.

**Comment by handrews** [comment]
@gregsdennis ok I lied I have one thing to say since I decided to reread and figure out why I thought this was done.

> The array is valid, not the element. The subschema must be applied to all elements when collecting annotations, but it does not say that the annotation applies to the element. These two combined implies that the annotations apply to the array.

The annotations from the subschema apply to the element(s) because the subschema applies to the element(s).  That's how annotations work.

The annotation from `contains`, if any, is associated with the array, just like the annotations from `items`/`prefixItems`.  These are all the same: they apply one or more subschemas to one or more array elements, providing validation and annotation results for the subschema + instance element.  They then combine the validation results (`contains` ORs them, the others AND them as it the usual behavior), and produce an annotation for the array as a whole indicating what parts of the array the application of the subschema covered.

This is why it never occurred to me that this was unresolved.  There's nothing strange about contains except that it ORs instead of ANDs.  But that's also true of `anyOf` vs `allOf` and we don't do anything strange with `anyOf` regarding annotations.

**Comment by handrews** [comment]
@Relequestual this needs a final call ASAP.

**Comment by gregsdennis** [comment]
@ssilverman and I have discussed this thoroughly in Slack and have come to the conclusion that `contains` for arrays is analogous to `required` for objects.

> ### [(Validation) 6.5.3. required](http://json-schema.org/draft/2019-09/json-schema-validation.html#rfc.section.6.5.3)
> The value of this keyword MUST be an array. Elements of this array, if any, MUST be strings, and MUST be unique.
> 
> An object instance is valid against this keyword if every item in the array is the name of a property in the instance.
> 
> Omitting this keyword has the same behavior as an empty array.

`required` considers the object as a whole.  Consider the case that this keyword only has a single property.  If the object does not contain that property, the entire object is marked as failed.  If the object does contain this property, then the _object_ acquires an annotation (as opposed to the property that was required).

> ### [(Core) 9.3.1.4. contains](http://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.9.3.1.4)
> The value of this keyword MUST be a valid JSON Schema.
> 
> _An array instance is valid against "contains" if at least one of its elements is valid against the given schema._ Note that when collecting annotations, the subschema MUST be applied to every array element even after the first match has been found. This is to ensure that all possible annotations are collected.

***NOTE** Italic text is unchanged since the keyword was introduced in Draft 6.*

Similarly `contains` considers the array as a whole.  If the array does not contain a matching item, the array is marked as failed.  If the array does contain a matching item, then the _array_ acquires an annotation, **not the matching item.**

The note that was added about annotations states that the `contains` subschema should be applied to all items to ensure all annotations are collected.  However, it does not state that those annotations should apply to the items.  The validation result applies to the array, not the items; therefore the annotations should also apply to the array, not the items.

> The annotations from the subschema apply to the element(s) because the subschema applies to the element(s). That's how annotations work.

Annotations should be applied as specified by the keyword that produces them.  `contains` is non-specific about annotation application, but it _is_ specific about where the validation result is applied, so annotations should be applied to the same target.

`unevaluatedProperties` does not consider `required`; and similarly `unevaluatedItems` should not consider `contains`.

If `unevaluatedItems` should consider `contains`, then I assert that `unevaluatedProperties` should consider `required`, which [it currently does not do](http://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.9.3.2.4).

Not only does this interpretation yield symmetry between objects and arrays, it matches more closely to the definition of the word "contains" and how the schema authors (not spec authors; we're a strange lot) would interpret that keyword _and_ the definition that the spec currently gives for it.

---

As a consequence of the realized parallelism with `required`, this opens a door to `contains` possibly supporting an array form where *all* subschemas must have matches within the array.  But this is best discussed in another issue.

**Comment by notEthan** [comment]
I don't think `required` is a good parallel to `contains`. it doesn't evaluate any subschema, which seems kind of central to `unevaluatedItems`. `required` is more parallel to `minItems` - they both simply assert the presence of child-instances with given property names / array indices.

I don't think there is a parallel keyword to `contains` which applies to objects.

**Comment by gregsdennis** [comment]
Had another discussion, and the point came up that single-form-`items` represents an AND operation (all items must match) whereas `contains` represents an OR operation (at least one must match).  In that respect, these also have a symmetry.

Building on that symmetry, `items` attaches annotations to the items in the array, so why shouldn't `contains`?

I'd be okay with applying annotations at _both_ the array and item level, given this.

**Comment by handrews** [comment]
@gregsdennis yes, `contains` is the OR version of `items`, otherwise they're identical (well, now except that `prefixItems` only prefixes `items`).

**Comment by notEthan** [comment]
I would like to clarify the instance location of the annotations of a child-schema applicator keyword - be it `items` or `contains` or `properties`. I thought this was clear from the spec, but it seems quite muddied from this discussion.

@handrews says (https://github.com/json-schema-org/json-schema-spec/issues/810#issuecomment-594080828):

> `contains` applies the subschema to each item in turn, and the annotations from that subschema are **most definitely attached _to the item_**. This is exactly the same as what `items` does.

and @gregsdennis says (https://github.com/json-schema-org/json-schema-spec/issues/810#issuecomment-631842268)
> `items` attaches annotations **to the items** in the array

(bolding added)

but the spec seems to me to indicate that the annotation of `items` is attached to the array, not to its (relevant) items.

[json-schema-core.html 9.3.1](https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.9.3.1)
> This keyword produces an annotation value which is the largest index to which this keyword applied a subschema.

it doesn't quite explicitly _say_ this is attached to the array, but the section on [collecting annotations](https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.7.7.1) indicates the instance location of an annotation is the one to which the keyword applies. for items that's the array.

moreover, it doesn't make sense to apply the annotations from `items` or `prefixItems` or `properties` to individual items or object values. those annotations indicate **where in the array/object** a subschema has been evaluated. if a schema `properties: {foo: {}, bar: {}}` produces an annotation of `[foo, bar]` against some instance, that annotation doesn't make sense on foo or bar or both, it applies to the object.

same with array subschema keywords producing annotations indicating applied indices, whether it's using true or max index or an array of indices.

this is long-winded for a bit of a sidetrack, but I think it's worth getting everybody on the same page, given how much of the above discussion is hung up on where these annotations apply.
